### PR TITLE
[SAGE-722] - Add `optgroup` support to `select` components

### PR DESCRIPTION
**Pull request by goodwinchris**
## Description

- [x] Add `optgroup` to `select` component
  - [x] Ensure disabled state on `optgroup`
  - [x] Ensure support for both `optgroup` and `options`

## Screenshots
CleanShot 2022-08-05 at 15 41 22@2x
CleanShot 2022-08-05 at 15 41 37@2x

## Testing in `sage-lib`
[Rails](http://localhost:4000/pages/component/form_select?tab=preview)
[React](http://localhost:4100/?path=/docs/sage-select--select-with-optgroups)

## Testing in `kajabi-products`
(**LOW**) Update component logic to support addition of `optgroup` in select data. `option`s should still maintain full support 

## Related

[SAGE-722: SageFormSelect needs `optgroup` affordance ](https://kajabi.atlassian.net/browse/SAGE-722)

### CODE REVIEW

**Review by anechol — APPROVED** [review]
LGTM

File: packages/sage-react/lib/Select/Select.jsx [hunk 1]
```diff
@@ -38,29 +66,17 @@ export const Select = ({
...
-            optionLabel = option;
-            optionValue = option;
-          } else {
-            optionLabel = option.label;
-            optionValue = option.value;
-            optionDisabled = option.disabled;
-          }
+        if (!isEmpty(optGroups)) {
+          optGroups.map((group) => (
+            <optgroup label={group.label} disabled={group.disabled}>
```
Comment (monkeypox8) [inline packages/sage-react/lib/Select/Select.jsx:71]: Similar to the `option` loop, these will also need a `key`.

**Review by monkeypox8 — CHANGES_REQUESTED** [review]
Working great! The `key` in the `optgroup` loop, and possibly the if statement in the return are the only blockers from approval from what I can tell.

File: packages/sage-react/lib/Select/Select.jsx [hunk 2]
```diff
@@ -25,6 +25,55 @@ export const Select = ({
...
+      optionDisabled;
+      // optionSelected;
+    if (typeof option === 'string') {
+      optionLabel = option;
+      optionValue = option;
+    } else {
+      optionLabel = option.label;
+      optionValue = option.value;
+      optionDisabled = option.disabled;
+      // optionSelected = option.selected;
```
Comment (goodwinchris) [inline packages/sage-react/lib/Select/Select.jsx:40]: 🔥

File: packages/sage-react/lib/Select/Select.jsx [hunk 3]
```diff
@@ -25,6 +25,55 @@ export const Select = ({
     }
   );
 
+  const buildOption = (option, i) => {
+    let optionLabel,
+      optionValue,
+      optionDisabled;
+      // optionSelected;
```
Comment (goodwinchris) [inline packages/sage-react/lib/Select/Select.jsx:32]: 🔥

File: docs/lib/sage_rails/app/sage_components/sage_form_select.rb [hunk 4]
```diff
@@ -7,11 +7,11 @@ class SageFormSelect < SageComponent
     message: [:optional, String],
     multiple: [:optional, TrueClass],
     name: [:optional, String],
-    select_options: [:optional, [[{
-      text: String,
-      value: [:optional, NilClass, String],
+    select_options: [:optional, [[SageSchemas::FORM_SELECT_OPTION, SageSchemas::FORM_SELECT_OPTGROUP]]],
+    select_optgroups: [:optional, [[{
```
Comment (goodwinchris) [inline docs/lib/sage_rails/app/sage_components/sage_form_select.rb:11]: Look into clean up

File: docs/app/views/examples/components/form_select/_props.html.erb [hunk 5]
```diff
@@ -24,7 +24,7 @@
 </tr>
 <tr>
   <td><%= md('`select_options`') %></td>
-  <td><%= md('An array of items that will serve as the `<select>`\'s `<option>`s') %></td>
+  <td><%= md('An array of items that will serve as the `<select>`\'s `<option>`s and `<optgroup>`s.') %></td>
```
Comment (goodwinchris) [inline docs/app/views/examples/components/form_select/_props.html.erb:27]: Question for reviewers: Should I add more detail here about _how_ to structure this?

File: docs/app/views/examples/components/form_select/_props.html.erb [hunk 6]
```diff
@@ -24,13 +24,15 @@
...
   <td><%= md('```
 Array<{
   text: String,
   value: String (optional),
   disabled: Boolean (false),
-  selected: Boolean (false)
+  selected: Boolean (false),
+  group_label: String,
+  group_options: Array,
```
Comment (goodwinchris) [inline docs/app/views/examples/components/form_select/_props.html.erb:35]: Question for reviewers: Should this have information about what can be contained in this array?

**Review by monkeypox8 — APPROVED** [review]
👍🏼 Added one suggestion on documentation

**Review by QuintonJason — APPROVED** [review]
Good to go. There's the non-blocking comment regarding the `Array` representation in the docs but still good.

**Review by anechol — APPROVED** [review]
LGTM

File: docs/lib/sage_rails/app/views/sage_components/_sage_form_select.html.erb [hunk 7]
```diff
@@ -18,14 +18,33 @@
...
+            <%= option[:disabled] ? " disabled" : nil -%>
+          >
+            <% option[:group_options].each do |option| %>
+              <option
+                value="<%= option[:value] %>"
+                <%= option[:selected] && !option[:disabled] ? " selected=true" : nil -%>
+                <%= option[:disabled] ? " disabled" : nil -%>
+              >
+                <%= option[:text] %>
+              </option>
```
Comment (ju-Skinner) [inline docs/lib/sage_rails/app/views/sage_components/_sage_form_select.html.erb:35]: This is a bit of tricky one. We are duplicating the same code down below, but this is also in a view in our library. Would be ideal to have a function helper to build out the options to reduce the code duplicity.

File: packages/sage-react/lib/Select/Select.jsx [hunk 8]
```diff
@@ -25,6 +25,51 @@ export const Select = ({
...
+        disabled={optionDisabled}
+      >
+        {optionLabel}
+      </option>
+    );
+  };
+
+  const buildOptgroup = (optgroup, i) => {
+    const optgroupLabel = optgroup.group_label;
+    const optgroupDisabled = optgroup.disabled;
```
Comment (ju-Skinner) [inline packages/sage-react/lib/Select/Select.jsx:54]: do we need to set these values to a variable?

File: packages/sage-react/lib/Select/Select.jsx [hunk 9]
```diff
@@ -25,6 +25,51 @@ export const Select = ({
     }
   );
 
+  const buildOption = (option, i) => {
+    let optionLabel,
+      optionValue,
+      optionDisabled;
+    if (typeof option === 'string') {
```
Comment (ju-Skinner) [inline packages/sage-react/lib/Select/Select.jsx:32]: I would change this conditional check to something like 

```
if ( option.hasOwnProperty('label') ) {
  optionLabel = option.label;
  optionValue = option.value;
  optionDisabled = option.disabled;
}
else {
  optionLabel = option; 
  optionValue = option;
}

**Review by ju-Skinner — CHANGES_REQUESTED** [review]
See comments...Open to discuss and/or pair if you have questions.

File: docs/app/views/examples/components/form_select/_props.html.erb (hunk 5 above)
Comment (ju-Skinner) [reply docs/app/views/examples/components/form_select/_props.html.erb:27]: The more documentation you can provide for developers the better.

**Review by ju-Skinner — APPROVED** [review]
LGTM...great work

**Review by philschanely — APPROVED** [review]
Solid work here, @goodwinchris . 🏆

**Pull request merged**